### PR TITLE
SE: Maintenance: Allocation free hash code implementation for immutable dict

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6964 

See also #6983

### CODE REVIEW

**Comment by martin-strecker-sonarsource** [comment]
I also did a benchmark.net test of the two implementations (the "original" vs. the "optimized" implementation):

* Org = Original
* Opt = Optimized
* Empty/2/10/100: number of elements in the dictionary
* The dictionary is `ImmutableDictionary<ISymbol, SymbolicValue>` where `ISymbol` is a fake (empty interface with an empty implementing class) and `SymbolicValue` is the original implementation. The implementation of DictionaryContentHash is identical to the ones in the diff of the PR.

| Method             | Runtime            |          Mean | Allocated |
|--------------------|--------------------|--------------:|----------:|
| HashCode_Org_Empty | .NET 7.0           |     143.42 us |         - |
| HashCode_Opt_Empty | .NET 7.0           |      28.88 us |         - |
|                    |                    |               |           |
| HashCode_Org_2     | .NET 7.0           |   3,092.78 us | 1520002 B |
| HashCode_Opt_2     | .NET 7.0           |   2,513.88 us |       2 B |
|                    |                    |               |           |
| HashCode_Org_10    | .NET 7.0           |  10,031.45 us | 1520008 B |
| HashCode_Opt_10    | .NET 7.0           |   8,632.35 us |       8 B |
|                    |                    |               |           |
| HashCode_Org_100   | .NET 7.0           |  90,781.06 us | 1520084 B |
| HashCode_Opt_100   | .NET 7.0           |  78,309.97 us |      72 B |
|                    |                    |               |           |
| HashCode_Org_Empty | .NET Framework 4.8 |     281.56 us |         - |
| HashCode_Opt_Empty | .NET Framework 4.8 |      94.00 us |         - |
|                    |                    |               |           |
| HashCode_Org_2     | .NET Framework 4.8 |   5,215.75 us | 1524523 B |
| HashCode_Opt_2     | .NET Framework 4.8 |   4,768.55 us |         - |
|                    |                    |               |           |
| HashCode_Org_10    | .NET Framework 4.8 |  16,837.61 us | 1524518 B |
| HashCode_Opt_10    | .NET Framework 4.8 |  16,006.34 us |         - |
|                    |                    |               |           |
| HashCode_Org_100   | .NET Framework 4.8 | 147,512.89 us | 1525760 B |
| HashCode_Opt_100   | .NET Framework 4.8 | 139,539.74 us |         - |

File: analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs [hunk 1]
```diff
@@ -30,8 +30,20 @@ public static class HashCode    // Replacement for System.HashCode that is avail
         private const int RotateOffset = 17;
         private const int IntSeed = 393241;
 
-        public static int DictionaryContentHash<TKey, TValue>(IDictionary<TKey, TValue> dictionary) =>
-            dictionary.Aggregate(0, (seed, kvp) => seed ^ Combine(kvp.Key, kvp.Value));
+        public static int DictionaryContentHash<TKey, TValue>(ImmutableDictionary<TKey, TValue> dictionary)
+        {
+            if (dictionary.IsEmpty)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/HashCode.cs:35]: Comment about performance please, so we don't "improve" it :D

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/HashCodeTest.cs [hunk 2]
```diff
@@ -43,17 +43,6 @@ public void Combine_ProducesDifferentResults(string input)
             hash5.Should().NotBe(0).And.NotBe(hash2).And.NotBe(hash3).And.NotBe(hash4);
         }
 
-        [TestMethod]
-        public void DictionaryContentHash_StableForUnsortedDictionary()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Helpers/HashCodeTest.cs:47]: Why was this removed?

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Just a questin about the UT

File: analyzers/tests/SonarAnalyzer.UnitTest/Helpers/HashCodeTest.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Helpers/HashCodeTest.cs:47]: This was for `System.Collection.Generic.Dictionary<,>` which isn't supported anymore. I tried to support both types but it wasn't possible with one implementation and so I removed the test (I even tried `HashCode.DictionaryContentHash<TDict, TKey, TValue>(TDict dict) where TDict: IDictionary<TKey, TValue>` but this still used `IDictionary.GetEnumerator()` for the call `DictionaryContentHash<ImmutableDictionary<int, int>, int, int>()` under the hood).

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
